src/net/dns.rs

```diff
@@ -42,33 +82,105 @@ pub(crate) async fn prune_dns_cache(context: &Context) -> Result<()> {
     Ok(())
 }
 
-/// Looks up the hostname and updates DNS cache
-/// on success.
+/// Map from hostname to IP addresses.
+static LOOKUP_HOST_CACHE: Lazy<std::sync::Mutex<HashMap<String, Vec<IpAddr>>>> =
+    Lazy::new(Default::default);
+
+async fn lookup_host_with_memory_cache(
+    context: &Context,
+    hostname: &str,
+    port: u16,
+) -> Result<Vec<IpAddr>> {
+    let stale_result = LOOKUP_HOST_CACHE.lock().unwrap().get(hostname).cloned();
+    if let Some(stale_result) = stale_result {
+        // Revalidate the cache in the background.
+        {
+            let context = context.clone();
+            let hostname = hostname.to_string();
+            let hostname_clone = hostname.clone();
+            tokio::spawn(async move {
+                match lookup_host((hostname.as_str(), port))
+                    .await
+                    .context("DNS lookup failure")
+                {
+                    Ok(res) => {
+                        let res = res.into_iter().map(|addr| addr.ip()).collect();
+                        LOOKUP_HOST_CACHE
+                            .lock()
+                            .unwrap()
+                            .insert(hostname_clone, res);
```

Comment:
Instead of introducing `hostname_clone`, this could just call `clone()` where a clone is needed:

```suggestion
            let hostname = hostname.to_string();
            tokio::spawn(async move {
                match lookup_host((hostname.clone(), port))
                    .await
                    .context("DNS lookup failure")
                {
                    Ok(res) => {
                        let res = res.into_iter().map(|addr| addr.ip()).collect();
                        LOOKUP_HOST_CACHE.lock().unwrap().insert(hostname, res);
```